### PR TITLE
fix: update sample region tag to parse host instead of proto package

### PR DESCRIPTION
**Pull request by alicejli**
Per SnippetGen design doc- the ApiShortName in the RegionTag should be generated from the host of the proto. 

Prior to this change, the Java generator parses the proto package name for the ApiShortName, leading to occasional discrepancies.

### CODE REVIEW

File: src/test/java/com/google/api/generator/gapic/composer/ComposerTest.java
Comment (diego92sigma6) [inline src/test/java/com/google/api/generator/gapic/composer/ComposerTest.java:45]: this non-constant class field should use lowerCamelCase

File: src/main/java/com/google/api/generator/gapic/composer/Composer.java
Comment (diego92sigma6) [inline src/main/java/com/google/api/generator/gapic/composer/Composer.java:33]: Seems like the IDE defaulted to use a star for a bunch of changed files

**Comment by alicejli** [comment]
Adding a `do not merge` label as I confirmed with Amanda that the `iam` service ApiShortName should be fixed (it currently reads as `api` because the hostname is `iam-meta-api.googleapis.com`. Will wait until a decision is made on how to handle this exception before fixing and merging.

File: src/test/java/com/google/api/generator/gapic/composer/ComposerTest.java
Comment (alicejli) [reply src/test/java/com/google/api/generator/gapic/composer/ComposerTest.java:45]: thanks! just updated.

File: src/main/java/com/google/api/generator/gapic/composer/Composer.java
Comment (alicejli) [reply src/main/java/com/google/api/generator/gapic/composer/Composer.java]: Thanks for catching! I reverted.

File: src/main/java/com/google/api/generator/gapic/composer/Composer.java
Comment (blakeli0) [inline src/main/java/com/google/api/generator/gapic/composer/Composer.java:235]: Try to use `StringSplitter` from Guava instead of the default String split, see https://errorprone.info/bugpattern/StringSplitter

File: src/main/java/com/google/api/generator/gapic/composer/Composer.java
Comment (blakeli0) [inline src/main/java/com/google/api/generator/gapic/composer/Composer.java:233]: I think the solution in C# is much cleaner, we can do similar things i n Java as well. We don't have to declare each pattern as a separate variable, and we don't have to have a `apiShortNameHelper` as well. Something like
```
String apiShortName = split using Splitter with . and get the first one
apiShortName = split using Splitter with - and get the last one
```
Usually it's a good practice to check the length before we get it to prevent runtime exception, but we don't have to check it here since a well formatted defaultHost should not error out.

File: src/main/java/com/google/api/generator/gapic/composer/samplecode/SettingsSampleComposer.java
Comment (blakeli0) [inline src/main/java/com/google/api/generator/gapic/composer/samplecode/SettingsSampleComposer.java:148]: Do we still need this TODO?

File: src/main/java/com/google/api/generator/gapic/model/GapicClass.java
Comment (blakeli0) [inline src/main/java/com/google/api/generator/gapic/model/GapicClass.java:49]: It's better to have less overloaded factory methods since it loses the benefit of having builders. See [here](go/java-practices/builders#when) for details. I would remove these two methods and use builders to add `defaultHost`. We should also examine all the places that created `GapicClass` and see if we really need to add `defaultHost` every time we create `GapicClass`.

File: src/main/java/com/google/api/generator/gapic/composer/Composer.java
Comment (blakeli0) [inline src/main/java/com/google/api/generator/gapic/composer/Composer.java:206]: Nice to have: I know this is pre-existing code, it would be nice to refactor this part to  be more readable/understandable

- `map` is usually used to map an object to another, here we are not really mapping, we are adding something to the existing object, so we can use `foreach` or just traditional for loop. 
- Nested mappings and lambda expressions make it really hard to read, there are six levels of indentation here, it's usually a code smell that you have more than two levels of indentation.

File: src/main/java/com/google/api/generator/gapic/composer/samplecode/SettingsSampleComposer.java
Comment (alicejli) [reply src/main/java/com/google/api/generator/gapic/composer/samplecode/SettingsSampleComposer.java:148]: Yes - this comment in particular is in reference to https://buganizer.corp.google.com/issues/239821325 which I decided to keep as a separate PR from this b/c just parsing the host got more complex than I initially realized.

File: src/main/java/com/google/api/generator/gapic/composer/Composer.java
Comment (alicejli) [reply src/main/java/com/google/api/generator/gapic/composer/Composer.java]: Thanks for the tip! Makes sense; updated.

File: src/main/java/com/google/api/generator/gapic/composer/Composer.java
Comment (alicejli) [reply src/main/java/com/google/api/generator/gapic/composer/Composer.java:232]: Thank you - the Guava splitter makes it much easier to mimic C#. Updated!

File: src/main/java/com/google/api/generator/gapic/model/GapicClass.java
Comment (alicejli) [reply src/main/java/com/google/api/generator/gapic/model/GapicClass.java]: Thank you; that makes sense. I've gone through and updated to use the Builder as well as removed it where we don't need to add the `defaultHost`.

File: src/test/java/com/google/api/generator/gapic/composer/ComposerTest.java
Comment (amanda-tarafa) [inline src/test/java/com/google/api/generator/gapic/composer/ComposerTest.java:141]: Hmmm, we don't capitalize the first letter of the API short name. Is this just for output or is it literally being used like this on the region tags?

File: src/test/java/com/google/api/generator/gapic/composer/ComposerTest.java
Comment (blakeli0) [inline src/test/java/com/google/api/generator/gapic/composer/ComposerTest.java:17]: Can you please update Intellij to prevent it from importing *?

File: src/test/java/com/google/api/generator/gapic/composer/ComposerTest.java
Comment (blakeli0) [inline src/test/java/com/google/api/generator/gapic/composer/ComposerTest.java:86]: We usually follow this [pattern](go/java-style#s5.2.3-method-names) for unit test names

File: src/main/java/com/google/api/generator/gapic/composer/Composer.java
Comment (alicejli) [reply src/main/java/com/google/api/generator/gapic/composer/Composer.java]: Thanks! Made an exception for the iam service.

File: src/main/java/com/google/api/generator/gapic/composer/Composer.java
Comment (alicejli) [reply src/main/java/com/google/api/generator/gapic/composer/Composer.java]: Thanks for the advice - makes sense. I've refactored to use `forEach`.

File: test/integration/goldens/asset/samples/snippets/generated/main/java/com/google/cloud/asset/v1/assetserviceclient/analyzeiampolicy/AsyncAnalyzeIamPolicy.java
Comment (alicejli) [reply test/integration/goldens/asset/samples/snippets/generated/main/java/com/google/cloud/asset/v1/assetserviceclient/analyzeiampolicy/AsyncAnalyzeIamPolicy.java:19]: Thanks! That makes sense to me; I dug into this a bit- the region tag builder is sanitizing the attributes to use `JavaStyle.toUpperCamelCase()` which is then getting converted to all lowercase in the actual sample, so changing the current setup is a little more involved than I thought. 

I will make a TODO to refactor this in a different PR.

File: src/test/java/com/google/api/generator/gapic/composer/ComposerTest.java
Comment (alicejli) [reply src/test/java/com/google/api/generator/gapic/composer/ComposerTest.java]: Thanks - I've updated my IntelliJ settings so this shouldn't happen anymore.

File: src/test/java/com/google/api/generator/gapic/composer/ComposerTest.java
Comment (alicejli) [reply src/test/java/com/google/api/generator/gapic/composer/ComposerTest.java]: Thanks; updated.

File: test/integration/goldens/kms/samples/snippets/generated/main/java/com/google/cloud/kms/v1/keymanagementserviceclient/updatecryptokeyversion/SyncUpdateCryptoKeyVersion.java
Comment (alicejli) [reply test/integration/goldens/kms/samples/snippets/generated/main/java/com/google/cloud/kms/v1/keymanagementserviceclient/updatecryptokeyversion/SyncUpdateCryptoKeyVersion.java:19]: Thanks Amanda! Yep I'm planning on fixing that here (still in draft): https://github.com/googleapis/gapic-generator-java/pull/1047 - can you confirm if the capitalization of the service / RPC name is important? Or is it okay to leave it as all lowercase?

Comment (alicejli) [reply test/integration/goldens/kms/samples/snippets/generated/main/java/com/google/cloud/kms/v1/keymanagementserviceclient/updatecryptokeyversion/SyncUpdateCryptoKeyVersion.java:19]: Confirmed that capitalization is important so will plan to fix that in the other PR as well.

File: src/test/java/com/google/api/generator/gapic/composer/ComposerTest.java
Comment (alicejli) [reply src/test/java/com/google/api/generator/gapic/composer/ComposerTest.java:147]: Answered below :)

**Comment by alicejli** [comment]
https://github.com/googleapis/gapic-generator-java/issues/1051

File: test/integration/goldens/kms/samples/snippets/generated/main/java/com/google/cloud/kms/v1/keymanagementservicesettings/getkeyring/SyncGetKeyRing.java
Comment (alicejli) [reply test/integration/goldens/kms/samples/snippets/generated/main/java/com/google/cloud/kms/v1/keymanagementservicesettings/getkeyring/SyncGetKeyRing.java:19]: Ah thanks for clarifying - the generator was already set up to generate these so I'll plan to keep for now, but good to know that these are optional.

**Review by amanda-tarafa — APPROVED** [review]
LGTM from the perspective of SnippetGen output and counting on #1047 to come after.
Thanks!

File: src/test/java/com/google/api/generator/gapic/composer/ComposerTest.java
Comment (blakeli0) [inline src/test/java/com/google/api/generator/gapic/composer/ComposerTest.java:92]: It's usually better to create three separate test cases for this scenario, in case one of them failed, it's easier for us to identify the issue. The test name can be something like `parseDefaultHost_shouldReturnApiShortNameIfHostContainsDash`, also can you please add a test case for the hardcode `iam`?

File: src/test/java/com/google/api/generator/gapic/composer/ComposerTest.java
Comment (blakeli0) [inline src/test/java/com/google/api/generator/gapic/composer/ComposerTest.java:103]: Can we just use the static imported `assertEquals`?

File: src/main/java/com/google/api/generator/gapic/composer/grpc/MockServiceClassComposer.java
Comment (blakeli0) [inline src/main/java/com/google/api/generator/gapic/composer/grpc/MockServiceClassComposer.java:72]: I feel like we might be adding `defaulHost` to too many composers, I'm pretty sure we don't need it here and `MockServiceImplClassComposer`, probably not in `ResourceNameHelperClassComposer.java` as well, can you try to run all the tests without adding defaultHost to these composers? It will make the tests much easier

File: src/test/java/com/google/api/generator/gapic/composer/grpc/GrpcServiceCallableFactoryClassComposerTest.java
Comment (blakeli0) [inline src/test/java/com/google/api/generator/gapic/composer/grpc/GrpcServiceCallableFactoryClassComposerTest.java:33]: I don't think we need the assertions in this tests, because we didn't touch `AbstractServiceCallableFactoryClassComposer` or `GrpcServiceCallableFactoryClassComposer`.

File: src/test/java/com/google/api/generator/gapic/composer/grpc/GrpcServiceStubClassComposerTest.java
Comment (blakeli0) [inline src/test/java/com/google/api/generator/gapic/composer/grpc/GrpcServiceStubClassComposerTest.java:32]: I don't think we need the assertions in this tests, because we didn't touch `GrpcServiceStubClassComposer` or `AbstractTransportServiceStubClassComposer`.

File: src/test/java/com/google/api/generator/gapic/composer/grpc/MockServiceImplClassComposerTest.java
Comment (blakeli0) [inline src/test/java/com/google/api/generator/gapic/composer/grpc/MockServiceImplClassComposerTest.java:33]: We don't need these tests if we can remove the `withDefaultHost` part from `MockServiceImplClassComposer`.

File: src/test/java/com/google/api/generator/gapic/composer/grpc/ServiceClientTestClassComposerTest.java
Comment (blakeli0) [inline src/test/java/com/google/api/generator/gapic/composer/grpc/ServiceClientTestClassComposerTest.java:33]: We don't need to touch this test since we didn't touch `ServiceClientTestClassComposer` or `AbstractServiceClientTestClassComposer`

File: src/test/java/com/google/api/generator/gapic/composer/resourcename/ResourceNameHelperClassComposerTest.java
Comment (blakeli0) [inline src/test/java/com/google/api/generator/gapic/composer/resourcename/ResourceNameHelperClassComposerTest.java:113]: We can remove these new assertions if we found `withDefaultHost` is not needed for `ResourceNameHelperClassComposerTest`

File: src/test/java/com/google/api/generator/gapic/composer/ComposerTest.java
Comment (alicejli) [reply src/test/java/com/google/api/generator/gapic/composer/ComposerTest.java]: Makes sense - updated!

File: src/test/java/com/google/api/generator/gapic/composer/ComposerTest.java
Comment (alicejli) [reply src/test/java/com/google/api/generator/gapic/composer/ComposerTest.java]: Yes good catch; updated

File: src/test/java/com/google/api/generator/gapic/composer/grpc/GrpcServiceCallableFactoryClassComposerTest.java
Comment (alicejli) [reply src/test/java/com/google/api/generator/gapic/composer/grpc/GrpcServiceCallableFactoryClassComposerTest.java]: Thanks that makes sense - I forgot to update after I refactored. Updated!

File: src/main/java/com/google/api/generator/gapic/composer/grpc/MockServiceClassComposer.java
Comment (alicejli) [reply src/main/java/com/google/api/generator/gapic/composer/grpc/MockServiceClassComposer.java]: That makes sense. Confirmed that removing `withDefaultHost` from the builders of `MockServiceClassComposer`, `MockServiceImplClassComposer`, and `ResourceNameHelperClassComposer` is fine as those composers are not used in generating samples. Have removed from builders as well as the unit tests.

File: src/test/java/com/google/api/generator/gapic/composer/grpc/MockServiceImplClassComposerTest.java
Comment (alicejli) [reply src/test/java/com/google/api/generator/gapic/composer/grpc/MockServiceImplClassComposerTest.java]: Confirmed; removed

File: src/test/java/com/google/api/generator/gapic/composer/resourcename/ResourceNameHelperClassComposerTest.java
Comment (alicejli) [reply src/test/java/com/google/api/generator/gapic/composer/resourcename/ResourceNameHelperClassComposerTest.java]: Confirmed, removed

File: src/test/java/com/google/api/generator/gapic/composer/grpc/ServiceClientTestClassComposerTest.java
Comment (alicejli) [reply src/test/java/com/google/api/generator/gapic/composer/grpc/ServiceClientTestClassComposerTest.java]: Removed

File: src/main/java/com/google/api/generator/gapic/model/GapicClass.java
Comment (blakeli0) [inline src/main/java/com/google/api/generator/gapic/model/GapicClass.java:75]: Looking at the PR again, I noticed that it's not the default host we are setting here, we are actually the apiShortName that are parsed from default host, do you mind renaming this field to `apiShortName`?

File: src/main/java/com/google/api/generator/gapic/model/HttpBindings.java
Comment (blakeli0) [inline src/main/java/com/google/api/generator/gapic/model/HttpBindings.java:49]: These code looks like from my another PR, you may need to merge from main again.

File: src/main/java/com/google/api/generator/gapic/model/GapicClass.java
Comment (blakeli0) [reply src/main/java/com/google/api/generator/gapic/model/GapicClass.java:75]: You are right, I was confused with what's parsed and what's being set in GapicClass. SpringCodeGen may want to reuse the parsing logic so I was thinking we might want to move the parsing logic earlier in the process, but that can be done in a different PR as well.

File: src/main/java/com/google/api/generator/gapic/model/HttpBindings.java
Comment (alicejli) [reply src/main/java/com/google/api/generator/gapic/model/HttpBindings.java]: Good catch - updated!

**Review by blakeli0 — APPROVED** [review]

**Pull request merged**
